File: CSEastAsianWidth/CharWidth.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ConsoleApplication1
{
    enum EastAsianWidths
    {
        Unknown = 0,

        FullWidth,
        HalfWidth,
        Wide,
        Narrow,
        Ambiguous,
        Neutral,

        //F = FullWidth,
        //H = HalfWidth,
        //W = Wide,
        //Na = Narrow,
        //A = Ambiguous,
        //N = Neutral
    }

    class Range
    {
        public Range(EastAsianWidths w, int code1, int code2)
        {
            this.Width = w;
            this.Code1 = code1;
            this.Code2 = code2;
        }
        public EastAsianWidths Width { get; private set; }
        public int Code1 { get; private set; }
        public int Code2 { get; private set; }

        public bool Contains(int code)
        {
            if (Code1 <= code && code <= Code2)
            {
                return true;
            }

            return false;
        }
        public bool Contains(char c)
        {
            return Contains((int)c);
        }
    }

    class EastAsianWidthDictionary
    {
        public static EastAsianWidthDictionary Default
        {
            get
            {
                if (_Default == null)
                {
                    _Default = EastAsianWidthDictionary.Create();
                }
                return _Default;
            }
        }
        private static EastAsianWidthDictionary _Default;

        private List<Range> List;

        private EastAsianWidthDictionary()
        {
            List = new List<Range>();
        }

        public static EastAsianWidthDictionary Create()
        {
            string fileName = "EastAsianWidth.txt";
            if (System.IO.File.Exists(fileName))
            {
                using (System.IO.StreamReader sr = new System.IO.StreamReader(fileName, Encoding.ASCII))
                {
                    return Create(sr);
                }
            }
            else
            {
                System.Net.WebClient client = new System.Net.WebClient();
                string s = client.DownloadString("http://www.unicode.org/Public/UCD/latest/ucd/EastAsianWidth.txt");
                using (System.IO.StringReader sr = new System.IO.StringReader(s))
                {
                    return Create(sr);
                }
            }
        }

        public static EastAsianWidthDictionary Create(System.IO.TextReader sr)
        {
            EastAsianWidthDictionary dic = new EastAsianWidthDictionary();

            System.Text.RegularExpressions.Regex reg
                = new System.Text.RegularExpressions.Regex(@"^(?<CODE1>[0-9A-F]+)(\.\.(?<CODE2>[0-9A-F]+))?;(?<WIDTH>(Na|N|W|F|H|A))\s");

            EastAsianWidthDictionary dictionary = new EastAsianWidthDictionary();

            string line;

            while (sr.Peek() != -1)
            {
                line = sr.ReadLine();
                if (line.StartsWith("#"))
                {
                    continue;
                }

                var match = reg.Match(line);
                if (match.Success)
                {
                    string sw = match.Groups["WIDTH"].Value;
                    EastAsianWidths w = EastAsianWidths.Unknown;
                    switch (sw[0])
                    {
                    case 'N': w = sw.Length == 1 ? EastAsianWidths.Neutral : EastAsianWidths.Narrow; break;
                    case 'W': w = EastAsianWidths.Wide; break;
                    case 'F': w = EastAsianWidths.FullWidth; break;
                    case 'H': w = EastAsianWidths.HalfWidth; break;
                    case 'A': w = EastAsianWidths.Ambiguous; break;
                    default: w = EastAsianWidths.Unknown; break;
                    }

                    int code1 = int.Parse(match.Groups["CODE1"].Value, System.Globalization.NumberStyles.HexNumber);
                    int code2 = code1;
                    if (match.Groups["CODE2"].Success)
                    {
                        code2 = int.Parse(match.Groups["CODE2"].Value, System.Globalization.NumberStyles.HexNumber);
                    }

                    Range r = new Range(w, code1, code2);

                    dic.List.Add(r);
                }
                else
                {
                }
            }
            return dic;
        }

        public EastAsianWidths GetEastAsianWidth(char c)
        {
            int code = (int)c;

            foreach (Range r in this.List)
            {
                if (r.Contains(code))
                {
                    return r.Width;
                }
            }
            return EastAsianWidths.Unknown;
        }


    }


    static class CharExtention
    {
        public static EastAsianWidths GetEastAsianWidth(this char c)
        {
            return EastAsianWidthDictionary.Default.GetEastAsianWidth(c);
        }

        public static bool IsWideEastAsianWidth(this char c)
        {
            var w = EastAsianWidthDictionary.Default.GetEastAsianWidth(c);
            return w == EastAsianWidths.Wide || w == EastAsianWidths.FullWidth;
        }
        public static bool IsFullWidthEastAsianWidth(this char c)
        {
            var w = EastAsianWidthDictionary.Default.GetEastAsianWidth(c);
            return w == EastAsianWidths.FullWidth;
        }

        public static bool IsNarrowEastAsianWidth(this char c)
        {
            var w = EastAsianWidthDictionary.Default.GetEastAsianWidth(c);
            return w == EastAsianWidths.Narrow || w == EastAsianWidths.HalfWidth;
        }

        public static bool IsHalfEastAsianWidth(this char c)
        {
            var w = EastAsianWidthDictionary.Default.GetEastAsianWidth(c);
            return w == EastAsianWidths.HalfWidth;
        }

        private static System.Text.Encoding sjis = System.Text.Encoding.GetEncoding("shift_JIS");

        public static EastAsianWidths GetEastAsianWidthFrom_SJIS(this char c)
        {
            byte[] bs = sjis.GetBytes(c.ToString());
            int byteCount = bs.Length;
            if (byteCount == 1)
            {
                if (bs[0] <= 0x1F || bs[0] == 0x7F)
                {
                    return EastAsianWidths.Neutral;
                }
                else if (0xA1 <= bs[0] && bs[0] <= 0xDF)
                {
                    return EastAsianWidths.HalfWidth;
                }
                else
                {
                    return EastAsianWidths.Narrow;
                }
            }
            else if (byteCount == 2)
            {
                return EastAsianWidths.Wide; //include FullWidth
            }
            return EastAsianWidths.Unknown;
        }
        public static bool IsWideEastAsianWidth_SJIS(this char c)
        {
            int byteCount = sjis.GetByteCount(c.ToString());
            return byteCount == 2;
        }


        public static EastAsianWidths GetEastAsianWidthFrom_VB6(this char c)
        {
            if ((int)c <= 0x1F || (int)c == 0x7F)
            {
                return EastAsianWidths.Neutral;
            }
            string s = c.ToString();
            if (s == Microsoft.VisualBasic.Strings.StrConv(s, Microsoft.VisualBasic.VbStrConv.Wide, 0x0411))
            {
                return EastAsianWidths.Wide;  //include FullWidth
            }
            else
            {
                return EastAsianWidths.Narrow; //include HalfWidth
            }
        }
        public static bool IsWideEastAsianWidth_VB6(this char c)
        {
            string s = c.ToString();
            return s == Microsoft.VisualBasic.Strings.StrConv(s, Microsoft.VisualBasic.VbStrConv.Wide, 0x0411);
        }
    }
}

File: VBEastAsianWidth/CharWidth.vb
'Microsoft Limited Public Licence

Enum EastAsianWidths

    Unknown = 0

    FullWidth
    HalfWidth
    Wide
    Narrow
    Ambiguous
    Neutral

    'F = FullWidth
    'H = HalfWidth
    'W = Wide
    'Na = Narrow
    'A = Ambiguous
    'N = Neutral
End Enum


Class Range
    Public Sub New(ByVal w As EastAsianWidths, ByVal code1 As Integer, ByVal code2 As Integer)
        Me._Width = w
        Me._Code1 = code1
        Me._Code2 = code2
    End Sub
    Public ReadOnly Property Width As EastAsianWidths
        Get
            Return _Width
        End Get
    End Property
    Private _Width As EastAsianWidths

    Public ReadOnly Property Code1 As Integer
        Get
            Return _Code1
        End Get
    End Property
    Private _Code1 As Integer

    Public ReadOnly Property Code2 As Integer
        Get
            Return _Code2
        End Get
    End Property
    Private _Code2 As Integer

    Public Function Contains(ByVal code As Integer) As Boolean

        If (Code1 <= code AndAlso code <= Code2) Then
            Return True
        End If

        Return False

    End Function

    Public Function Contains(ByVal c As Char) As Boolean
        Return Contains(AscW(c))
    End Function
End Class

Class EastAsianWidthDictionary

    Public Shared ReadOnly Property [Default] As EastAsianWidthDictionary

        Get
            If (_Default Is Nothing) Then
                _Default = EastAsianWidthDictionary.Create()
            End If
            Return _Default
        End Get
    End Property
    Private Shared _Default As EastAsianWidthDictionary

    Private list As List(Of Range)

    Private Sub New()
        Me.list = New List(Of Range)()
    End Sub

    Public Shared Function Create() As EastAsianWidthDictionary

        Dim fileName As String = "_EastAsianWidth.txt"
        If (System.IO.File.Exists(fileName)) Then
            Using sr As New System.IO.StreamReader(fileName, System.Text.Encoding.ASCII)
                Return Create(sr)
            End Using
        Else
            Dim client As New System.Net.WebClient()
            Dim s As String = client.DownloadString("http://www.unicode.org/Public/UCD/latest/ucd/EastAsianWidth.txt")
            Using sr = New System.IO.StringReader(s)
                Return Create(sr)
            End Using
            end if
    End Function

    Public Shared Function Create(ByVal sr As System.IO.TextReader) As EastAsianWidthDictionary

        Dim dic As New EastAsianWidthDictionary()

        Dim reg As New System.Text.RegularExpressions.Regex("^(?<CODE1>[0-9A-F]+)(\.\.(?<CODE2>[0-9A-F]+))?(?<WIDTH>(Na|N|W|F|H|A))\s")

        Dim Dictionary As New EastAsianWidthDictionary()

        Dim line As String

        Do While (sr.Peek() <> -1)

            line = sr.ReadLine()
            If (line.StartsWith("#")) Then
                Continue Do
            End If

            Dim match As System.Text.RegularExpressions.Match = reg.Match(line)
            If (match.Success) Then

                Dim sw As String = match.Groups("WIDTH").Value
                Dim w As EastAsianWidths = EastAsianWidths.Unknown
                Select Case (sw)

                    Case "N"
                        w = EastAsianWidths.Neutral
                    Case "Na"
                        w = EastAsianWidths.Narrow
                    Case "W"
                        w = EastAsianWidths.Wide
                    Case "F"
                        w = EastAsianWidths.FullWidth
                    Case "H"
                        w = EastAsianWidths.HalfWidth
                    Case "A"
                        w = EastAsianWidths.Ambiguous
                    Case Else
                        w = EastAsianWidths.Unknown
                End Select

                Dim code1 As Integer = Integer.Parse(match.Groups("CODE1").Value, System.Globalization.NumberStyles.HexNumber)
                Dim code2 As Integer = code1
                If (match.Groups("CODE2").Success) Then
                    code2 = Integer.Parse(match.Groups("CODE2").Value, System.Globalization.NumberStyles.HexNumber)
                End If

                Dim r As Range = New Range(w, code1, code2)

                dic.List.Add(r)
            End If

        Loop
        Return dic
    End Function

    Public Function GetEastAsianWidth(ByVal c As Char) As EastAsianWidths
        Dim code As Integer = AscW(c)
        For Each r As Range In Me.list
            If (r.Contains(code)) Then
                Return r.Width
            End If
        Next
        Return EastAsianWidths.Unknown
    End Function


End Class


Module CharExtention

    <System.Runtime.CompilerServices.Extension> _
    Public Function GetEastAsianWidth(ByVal c As Char) As EastAsianWidths
        Return EastAsianWidthDictionary.Default.GetEastAsianWidth(c)
    End Function

    <System.Runtime.CompilerServices.Extension> _
    Public Function IsWideEastAsianWidth(ByVal c As Char) As Boolean
        Dim w As EastAsianWidths = EastAsianWidthDictionary.Default.GetEastAsianWidth(c)
        Return w = EastAsianWidths.Wide OrElse w = EastAsianWidths.FullWidth
    End Function

    <System.Runtime.CompilerServices.Extension> _
    Public Function IsFullWidthEastAsianWidth(ByVal c As Char) As Boolean

        Dim w = EastAsianWidthDictionary.Default.GetEastAsianWidth(c)
        Return w = EastAsianWidths.FullWidth
    End Function

    <System.Runtime.CompilerServices.Extension> _
    Public Function IsNarrowEastAsianWidth(ByVal c As Char) As Boolean
        Dim w As EastAsianWidths = EastAsianWidthDictionary.Default.GetEastAsianWidth(c)
        Return w = EastAsianWidths.Narrow OrElse w = EastAsianWidths.HalfWidth
    End Function

    <System.Runtime.CompilerServices.Extension> _
    Public Function IsHalfEastAsianWidth(ByVal c As Char) As Boolean
        Dim w = EastAsianWidthDictionary.Default.GetEastAsianWidth(c)
        Return w = EastAsianWidths.HalfWidth
    End Function

    Private sjis As System.Text.Encoding = System.Text.Encoding.GetEncoding("shift_JIS")

    <System.Runtime.CompilerServices.Extension> _
    Public Function GetEastAsianWidthFrom_SJIS(ByVal c As Char) As EastAsianWidths

        Dim bs() As Byte = sjis.GetBytes(c.ToString())
        Dim byteCount As Integer = bs.Length
        If (byteCount = 1) Then
            If (bs(0) <= &H1F OrElse bs(0) = &H7F) Then
                Return EastAsianWidths.Neutral
            ElseIf (&HA1 <= bs(0) AndAlso bs(0) <= &HDF) Then
                Return EastAsianWidths.HalfWidth
            Else
                Return EastAsianWidths.Narrow
            End If

        ElseIf (byteCount = 2) Then
            Return EastAsianWidths.Wide 'include FullWidth
        End If
        Return EastAsianWidths.Unknown
    End Function

    <System.Runtime.CompilerServices.Extension> _
    Public Function IsWideEastAsianWidth_SJIS(ByVal c As Char) As Boolean

        Dim byteCount As Integer = sjis.GetByteCount(c.ToString())
        Return byteCount = 2

    End Function

    <System.Runtime.CompilerServices.Extension> _
    Public Function GetEastAsianWidthFrom_VB6(ByVal c As Char) As EastAsianWidths

        If (AscW(c) <= &H1F OrElse AscW(c) = &H7F) Then
            Return EastAsianWidths.Neutral
        End If

        Dim s As String = c.ToString()
        If (s = Microsoft.VisualBasic.Strings.StrConv(s, Microsoft.VisualBasic.VbStrConv.Wide, &H411)) Then
            Return EastAsianWidths.Wide  'include FullWidth
        Else

            Return EastAsianWidths.Narrow 'include HalfWidth
        End If
    End Function

    <System.Runtime.CompilerServices.Extension> _
    Public Function IsWideEastAsianWidth_VB6(ByVal c As Char) As Boolean
        Dim s As String = c.ToString()
        Return s = Microsoft.VisualBasic.Strings.StrConv(s, Microsoft.VisualBasic.VbStrConv.Wide, &H411)
    End Function
End Module